lib/simplecov/result_merger.rb

```diff
@@ -87,6 +94,25 @@ def store_result(result)
         end
         true
       end
+
+    private
+
+      # Ensure only one process is reading or writing the resultset at any
+      # given time
+      def synchronize_resultset
+        # make it reentrant
+        return yield if defined?(@resultset_locked) && @resultset_locked
```

Comment:
Yes, so the idea is that if we call `synchronize_resultset` when we are already inside a `synchronize_resultset` call, we just let it through, since we know we already have a lock from the outer call. So we're not circumventing the lock, so much as saying "we already have a lock higher in the call stack, so just run the block". Due to how flock works, we would get a deadlock if we didn't, since the inner flock would wait forever for the outer one to release.

The reason we need to do this is we want to make both reads and writes acquire a lock so that neither one does anything unsafe. And in order to safely do a write, we 1. first need to read and 2. we need to ensure nobody else writes between our read and our write, otherwise we might lose data. So to accomplish that, write does `synchronize_resultset` around the whole thing so that any other processes have to wait. And since its nested read will also do `synchronize_resultset`, we use `@resultset_locked` to let it know it doesn't actually need a lock since we already got one.